File: backend/src/Backend.Infrastructure/Xtream/LooseJson.cs
using System.Globalization;
using System.Text.Json;

namespace Backend.Infrastructure.Xtream;

/// <summary>
/// Readers that tolerate Xtream panels mixing strings, numbers, nulls and empty arrays for the same field.
/// </summary>
internal static class LooseJson
{
    public static JsonElement? Property(this JsonElement element, string name) =>
        element.ValueKind == JsonValueKind.Object && element.TryGetProperty(name, out var value) ? value : null;

    public static string? String(this JsonElement element, string name)
    {
        var value = element.Property(name);
        var text = value?.ValueKind switch
        {
            JsonValueKind.String => value.Value.GetString(),
            JsonValueKind.Number => value.Value.GetRawText(),
            JsonValueKind.True => "1",
            JsonValueKind.False => "0",
            _ => null,
        };
        return string.IsNullOrWhiteSpace(text) ? null : text.Trim();
    }

    public static int? Int(this JsonElement element, string name) =>
        double.TryParse(element.String(name), NumberStyles.Float, CultureInfo.InvariantCulture, out var number)
            ? (int)number
            : null;

    public static double? Double(this JsonElement element, string name) =>
        double.TryParse(element.String(name), NumberStyles.Float, CultureInfo.InvariantCulture, out var number)
            ? number
            : null;

    public static bool Bool(this JsonElement element, string name) =>
        element.String(name) is { } text && (text == "1" || text.Equals("true", StringComparison.OrdinalIgnoreCase));

    public static DateTimeOffset? UnixTime(this JsonElement element, string name) =>
        long.TryParse(element.String(name), NumberStyles.Integer, CultureInfo.InvariantCulture, out var seconds) && seconds > 0
            ? DateTimeOffset.FromUnixTimeSeconds(seconds)
            : null;

    /// <summary>Reads a string array, a single string, or nothing. Drops blanks.</summary>
    public static IReadOnlyList<string> StringList(this JsonElement element, string name)
    {
        var value = element.Property(name);
        return value?.ValueKind switch
        {
            JsonValueKind.Array => value.Value.EnumerateArray()
                .Where(item => item.ValueKind == JsonValueKind.String && !string.IsNullOrWhiteSpace(item.GetString()))
                .Select(item => item.GetString()!.Trim())
                .ToList(),
            JsonValueKind.String when !string.IsNullOrWhiteSpace(value.Value.GetString()) => [value.Value.GetString()!.Trim()],
            _ => [],
        };
    }

    public static IEnumerable<JsonElement> ArrayItems(this JsonElement element) =>
        element.ValueKind == JsonValueKind.Array ? element.EnumerateArray() : [];
}

File: packages/shared/src/direct/looseJson.ts
/** Readers that tolerate Xtream panels mixing strings, numbers, nulls and empty arrays for one field. Mirrors backend LooseJson. */
export type Json = unknown;

export const isObject = (value: Json): value is Record<string, Json> =>
  typeof value === 'object' && value !== null && !Array.isArray(value);

export const prop = (value: Json, name: string): Json => (isObject(value) ? value[name] : undefined);

export function str(value: Json, name: string): string | null {
  const field = prop(value, name);
  const text =
    typeof field === 'string' ? field : typeof field === 'number' ? String(field) : typeof field === 'boolean' ? (field ? '1' : '0') : null;
  const trimmed = text?.trim();
  return trimmed ? trimmed : null;
}

export function num(value: Json, name: string): number | null {
  const text = str(value, name);
  if (text === null) return null;
  const number = Number(text);
  return Number.isFinite(number) ? number : null;
}

export function int(value: Json, name: string): number | null {
  const number = num(value, name);
  return number === null ? null : Math.trunc(number);
}

export function bool(value: Json, name: string): boolean {
  const text = str(value, name);
  return text === '1' || text?.toLowerCase() === 'true';
}

/** Unix seconds → ISO string; null for missing, zero or non-integer values. */
export function unixTime(value: Json, name: string): string | null {
  const text = str(value, name);
  if (text === null || !/^\d+$/.test(text)) return null;
  const seconds = Number(text);
  return seconds > 0 ? new Date(seconds * 1000).toISOString() : null;
}

/** A string array, a single string, or nothing. Drops blanks. */
export function strList(value: Json, name: string): string[] {
  const field = prop(value, name);
  if (Array.isArray(field))
    return field.filter((item): item is string => typeof item === 'string' && item.trim() !== '').map((item) => item.trim());
  return typeof field === 'string' && field.trim() ? [field.trim()] : [];
}

export const items = (value: Json): Json[] => (Array.isArray(value) ? value : []);
